### PR TITLE
Spanner: CMEK samples

### LINKED ISSUE #783: spanner: CMEK samples
- create database
- create backup
- restore database

### PR DESCRIPTION
**Pull request by jiren**
## Description

- create database
- create backup
- restore database

Fixes #783

## Checklist
- [x] **Tests** pass
- [x] **Lint** pass: `bundle exec rubocop`
- [x] Please **merge** this PR for me once it is approved.

### CODE REVIEW

File: spanner/spanner_samples.rb [hunk 1]
```diff
@@ -1721,6 +1791,37 @@ def restore_backup project_id:, instance_id:, database_id:, backup_id:
...
+    kms_key_name: kms_key_name
+  }
+  job = backup.restore database_id, encryption_config: encryption_config
+
+  puts "Waiting for restore backup operation to complete"
+
+  job.wait_until_done!
+
+  restore_info = job.database.restore_info
+  puts "Database #{restore_info.backup_info.source_database_id} was restored to #{database_id} from backup #{restore_info.backup_info.backup_id} using encryption key #{kms_key_name}"
```
[the remark this answers is not in the document]
Comment (jiren) [reply spanner/spanner_samples.rb]: done

**Comment by skuruppu** [comment]
@jiren, the Ruby client library was released with CMEK support so if you could update this PR then I think the tests will now pass.

**Comment by jiren** [comment]
> @jiren, the Ruby client library was released with CMEK support so if you could update this PR then I think the tests will now pass.

updated spanner gem.

**Comment by jiren** [comment]
`GOOGLE_CLOUD_SPANNER_TEST_INSTANCE and/or GOOGLE_CLOUD_SPANNER_PROJECT` env variable not exported, look like issue with Kokoro config.

**Review by thiagotnunes — APPROVED** [review]

**Comment by larkee** [comment]
@dazuma  Could we get some help debugging these test failures? I think they have been failing for some time.

**Comment by skuruppu** [comment]
For Ruby 2.5 and Ruby 2.6, I'm pretty sure the Spanner tests don't run intentionally. But there is this failure:
```
+ bundle exec rubocop
Inspecting 516 files
.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................C............

Offenses:

spanner/spanner_samples.rb:132:47: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
  job = instance.create_database database_id, statements: [ ...
                                              ^^^^^^^^^^^^^
spanner/spanner_samples.rb:147:3: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
  encryption_config: { kms_key_name: kms_key_name }
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
spanner/spanner_samples.rb:1756:5: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
    kms_key_name: kms_key_name
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
spanner/spanner_samples.rb:1812:5: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
    kms_key_name: kms_key_name
    ^^^^^^^^^^^^^^^^^^^^^^^^^^

516 files inspected, 4 offenses detected
```

@jiren can you please fix it?

For Ruby 2.7, it fails because the project doesn't have permissions to encrypt/decrypt. I asked @dazuma offline to give the necessary permissions.

**Comment by jiren** [comment]
> For Ruby 2.5 and Ruby 2.6, I'm pretty sure the Spanner tests don't run intentionally. But there is this failure:
> 
> ```
> + bundle exec rubocop
> Inspecting 516 files
> .......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................C............
> 
> Offenses:
> 
> spanner/spanner_samples.rb:132:47: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
>   job = instance.create_database database_id, statements: [ ...
>                                               ^^^^^^^^^^^^^
> spanner/spanner_samples.rb:147:3: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
>   encryption_config: { kms_key_name: kms_key_name }
>   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> spanner/spanner_samples.rb:1756:5: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
>     kms_key_name: kms_key_name
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^
> spanner/spanner_samples.rb:1812:5: C: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.
>     kms_key_name: kms_key_name
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 516 files inspected, 4 offenses detected
> ```
> 
> @jiren can you please fix it?
> 
> For Ruby 2.7, it fails because the project doesn't have permissions to encrypt/decrypt. I asked @dazuma offline to give the necessary permissions.

Done

**Comment by skuruppu** [comment]
Thanks @jiren, the 2.5 and 2.6 builds pass now :)

**Comment by skuruppu** [comment]
Phew! The tests have the right permissions now. One more error though:

```
Google::Cloud::FailedPreconditionError:
  9:KMS key provided can not protect the Spanner resource: Incompatible CryptoKey location. Resource location: us-east1, Key location: us-central1.. debug_error_string:{"created":"@1617839743.454502325","description":"Error received from peer ipv4:172.217.203.95:443","file":"src/core/lib/surface/call.cc","file_line":1067,"grpc_message":"KMS key provided can not protect the Spanner resource: Incompatible CryptoKey location. Resource location: us-east1, Key location: us-central1.","grpc_status":9}
```

@jiren the key and the instance need to be in the same region. If you can look into configuring the tests to reflect this, that would be great.

**Comment by jiren** [comment]
@skuruppu @thiagotnunes  Still facing KMS permission issue.

**Comment by hengfengli** [comment]
@jiren Can you please update the status of this PR? It is still a `Draft`, not an open PR.

**Comment by jiren** [comment]
> @jiren Can you please update the status of this PR? It is still a `Draft`, not an open PR.

done

**Comment by larkee** [comment]
The tests are still failing because the KMS key is in a different location from the instance.

The instance is in [`us-central`](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/226bdb759a3928fc13dfc6a1d8ddaf2e0e07723d/spanner/spanner_samples.rb#L27) but the KMS key is currently being created in `us-east`. Please update the key to use the correct location.

**Comment by jiren** [comment]
> The tests are still failing because the KMS key is in a different location from the instance.
> 
> The instance is in [`us-central`](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/226bdb759a3928fc13dfc6a1d8ddaf2e0e07723d/spanner/spanner_samples.rb#L27) but the KMS key is currently being created in `us-east`. Please update the key to use the correct location.

@larkee before kms zone was us-central-1 but it was not working. 
Ref: https://github.com/GoogleCloudPlatform/ruby-docs-samples/pull/784/commits/74ff7a2a8d9a4da0640ab5b843536a11c044098f

**Comment by larkee** [comment]
> > The tests are still failing because the KMS key is in a different location from the instance.
> > The instance is in [`us-central`](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/226bdb759a3928fc13dfc6a1d8ddaf2e0e07723d/spanner/spanner_samples.rb#L27) but the KMS key is currently being created in `us-east`. Please update the key to use the correct location.
> 
> @larkee before kms zone was us-central-1 but it was not working.
> Ref: [74ff7a2](https://github.com/GoogleCloudPlatform/ruby-docs-samples/commit/74ff7a2a8d9a4da0640ab5b843536a11c044098f)

Ahhh, try `us-central1`. 
That's what Python uses: see [here](https://github.com/googleapis/python-spanner/pull/275/files#diff-68f17b68549041db75595a9f359280586209fb34899e44b65634298465601f35R69)

**Comment by jiren** [comment]
> > > The tests are still failing because the KMS key is in a different location from the instance.
> > > The instance is in [`us-central`](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/226bdb759a3928fc13dfc6a1d8ddaf2e0e07723d/spanner/spanner_samples.rb#L27) but the KMS key is currently being created in `us-east`. Please update the key to use the correct location.
> > 
> > 
> > @larkee before kms zone was us-central-1 but it was not working.
> > Ref: [74ff7a2](https://github.com/GoogleCloudPlatform/ruby-docs-samples/commit/74ff7a2a8d9a4da0640ab5b843536a11c044098f)
> 
> Ahhh, try `us-central1`.
> That's what Python uses: see [here](https://github.com/googleapis/python-spanner/pull/275/files#diff-68f17b68549041db75595a9f359280586209fb34899e44b65634298465601f35R69)

Tried `us-central1` and failed with the permission issue.  This issue because of `GOOGLE_CLOUD_SPANNER_TEST_INSTANCE ` instance is in 'us-east1' and test KMS key in `us-central1`. This will be fixed only if KMS key created in `us-east1` or move spanner test instance `us-central1`  https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/da658aeadb439ce1f6286b57c9e3e0dcc76816d9/spanner/spec/spec_helper.rb#L25

**Comment by jiren** [comment]
@larkee @skuruppu tests are still failing due to KMS permission issue.

**Review by larkee — APPROVED** [review]

**Pull request merged**